packages/apollo-server-core/src/runHttpQuery.ts

```diff
@@ -132,31 +157,57 @@ export async function runHttpQuery(
         }
       }
 
-      if (
-        queryString === undefined &&
-        extensions &&
-        extensions.persistedQuery
-      ) {
-        // It looks like we've received an Apollo Persisted Query. Apollo Server
-        // does not support persisted queries out of the box, so we should fail
-        // fast with a clear error saying that we don't support APQs. (A future
-        // version of Apollo Server may support APQs directly.)
-        throw new HttpQueryError(
+      if (extensions && extensions.persistedQuery) {
+        // It looks like we've received an Apollo Persisted Query. Check if we
+        // support them. In an ideal world, we always would, however since the
+        // middleware options are created every request, it does not make sense
+        // to create a default cache here and save a referrence to use across
+        // requests
+        if (
+          !optionsObject.persistedQueries ||
+          !optionsObject.persistedQueries.cache
+        ) {
           // Return 200 to simplify processing: we want this to be intepreted by
           // the client as data worth interpreting, not an error.
-          200,
-          JSON.stringify({
-            errors: [
-              {
-                message: 'PersistedQueryNotSupported',
-              },
-            ],
-          }),
-          true,
-          {
-            'Content-Type': 'application/json',
-          },
-        );
+          throwHttpGraphQLError(
+            200,
+            [new PersistedQueryNotSupportedError()],
+            optionsObject,
+          );
+        } else if (extensions.persistedQuery.version !== 1) {
+          throw new HttpQueryError(400, 'Unsuported persited query version');
+        }
+
+        const sha = extensions.persistedQuery.sha256Hash;
+
+        if (queryString === undefined) {
+          queryString = await optionsObject.persistedQueries.cache.get(sha);
+          if (!queryString) {
+            throwHttpGraphQLError(
+              200,
+              [new PersistedQueryNotFoundError()],
+              optionsObject,
+            );
+          }
+        } else {
+          const calculatedSha = sha256()
+            .update(queryString)
+            .digest('hex');
+          if (sha !== calculatedSha) {
+            throw new HttpQueryError(400, 'provided sha does not match query');
+          }
+
+          optionsObject.persistedQueries.cache
```

Comment:
worth having a comment explicitly saying we're purposefully not waiting here.

also you could do the *whole* operation async by starting with Promise.resolve().then, but maybe this is fine.